modules/tinymce/src/plugins/directionality/main/ts/core/Direction.ts

```diff
@@ -5,28 +5,43 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
-import Editor from 'tinymce/core/api/Editor';
-import Tools from 'tinymce/core/api/util/Tools';
+import { Arr } from '@ephox/katamari';
+import { Traverse, Attribute, SugarElement } from '@ephox/sugar';
 
-const setDir = (editor: Editor, dir: string) => {
-  const dom = editor.dom;
-  let curDir;
-  const blocks = editor.selection.getSelectedBlocks();
+import Editor from 'tinymce/core/api/Editor';
 
-  if (blocks.length) {
-    curDir = dom.getAttrib(blocks[0], 'dir');
+type Dir = 'rtl' | 'ltl';
 
-    Tools.each(blocks, (block) => {
-      // Add dir to block if the parent block doesn't already have that dir
-      if (!dom.getParent(block.parentNode, '*[dir="' + dir + '"]', dom.getRoot())) {
-        dom.setAttrib(block, 'dir', curDir !== dir ? dir : null);
+const setDir = (editor: Editor, dir: Dir) => {
+  const selectedBlocks = editor.selection.getSelectedBlocks();
+  Arr.each(selectedBlocks, (block) => {
+    const sugarBlock = SugarElement.fromDom(block);
+    const blockParent = Traverse.parent(sugarBlock);
+    blockParent.map((blockParent: SugarElement<Element>) => {
+      const blockParentDirection = Attribute.get(blockParent, 'dir');
+      if ((blockParentDirection === undefined) || (blockParentDirection !== dir)) {
+        setDirAttr(editor, sugarBlock, dir);
+      } else { // if parent and child dir are going to be the same then remove it from child
+        Attribute.remove(sugarBlock, 'dir');
+        editor.nodeChanged();
       }
     });
+  });
+};
 
-    editor.nodeChanged();
+const setDirAttr = (editor: Editor, element: SugarElement, dir: Dir): void => {
```

Comment:
```suggestion
const setDirAttr = (editor: Editor, element: SugarElement<Element>, dir: Dir): void => {
```
Missing type here for what the sugar element contains.

Reply:
Fixed 👍